README.md
# Brig

A safe, scalable way to queue untrusted code to be run.

## How to use

TODO

## WIP

Do not use

CHANGELOG.md
# Changes

0.0.1 - WIP
- TODO

test.js
const {Brig} = require('.');

// TODO: replace hardcoded test

const brig = new Brig();

brig
    .open()
    .send('Math.random() + 1', {msg: 'hello world'}, response)
    .send('brig.context.msg', {msg: 'hello world'}, response)
    .send('brig.context.msg;', {msg: '2nd'}, response);

function response(err, resp) {
    console.log(resp);
}

worker.js
// list of valid module names
const MODULES = ['workerpool', 'debug', 'safe-buffer', 'stream'];

function require_worker(module) {
    if (!(MODULES.includes(module))) {
        throw new Error(`Invalid module: ${module}`);
    }
    return require(module);
}

// make our requires for the worker
const workerpool = require_worker('workerpool');
const {Buffer} = require_worker('safe-buffer');
const debug = require_worker('debug')('brig:worker');
// TODO: dynamic require based one main thread message ???
const {BrigRunnerVM2} = require('.');

const VM_OPTIONS = {
    timeout: process.env.WORKER_TIMEOUT === 'Infinity' ? undefined : 1000 * 60,
    require: {
        builtin: ['stream', 'http2'],
    }
};

let runner = new BrigRunnerVM2(VM_OPTIONS);

// monkey patch require to noop func
require, require_worker = () => {};

function workerExecute(code, context) {
    debug('workerExecute(...)');

    return runner.run(code, context);
}

function workerCompile(code) {
    debug('workerCompile(...)');

    return runner.compile(code);
}

workerpool.worker({
    execute: workerExecute,
    compile: workerCompile,
});

BrigRunner.js
const debug = require('debug')('brig:runner');

class BrigRunner {

    #options = null;
    #runner = null;

    constructor(options) {
        debug('BrigRunner(...)');
        this.#options = options;

        // init runner
        // runner = new VM(options);
    }

    compile(code) {
        debug('BrigRunner compile(...)');

        throw new Error('Must extend BrigRunner override compile(code)!');
    }

    wrap(code) {
        debug('BrigRunner wrap(...)');

        // Example code wrapper

        const wrapped = `
        require, eval = () => { throw new Error('Function not permitted!') };
        {
            (() => {
                return (${code});
            }).call(null);
        }
        `;

        return wrapped;
    }

    run(code, context) {
        debug('BrigRunner run(...)');

        throw new Error('Must extend BrigRunner override run(code)!');
    }

}

module.exports = BrigRunner;

BrigRunnerVM2.js
const debug = require('debug')('brig:runner');
const BrigRunner = require('./BrigRunner.js');
const {VM, VMScript} = require('vm2');

class BrigRunnerVM2 extends BrigRunner {

    #options = {
        timeout: process.env.BRIG_TIMEOUT === 'Infinity' ? undefined : 1000 * 60,
        fixAsync: true,
        eval: false,
        wasm: false,
    };
    #runner = null;

    constructor(options) {
        debug('BrigRunnerVM2(...)');
        super(options);
        if (options) this.#options = options;

        // init runner
        this.#runner = this.#initRunner(options);

        debug('BrigRunnerVM2 constructed');
    }

    #initRunner(opt) {
        const runner = new VM(opt);
        return runner;
    }

    compile(code) {
        debug('BrigRunnerVM2 compile(...)');

        const script = new VMScript(code);

        debug('BrigRunnerVM2 compiled');
        return script;
    }

    wrap(code) {
        debug('BrigRunnerVM2 wrap(...)');

        // Example code wrapper

        const wrapped = `
        ((context) => {
            return ${code};
        })(brig || { context: {}, date: new Date(), brig: true });
        `;

        debug('BrigRunnerVM2 wrapped');
        return wrapped;
    }

    run(code, context) {
        debug('BrigRunnerVM2 run(...)');

        code = this.wrap(code);
        if (context) this.#runner.freeze({ context, date: new Date(), brig: true }, 'brig');
        code = this.compile(code);
        const ret = this.#runner.run(code);

        debug('BrigRunnerVM2 ran');
        return ret;
    }

}

module.exports = BrigRunnerVM2;

brig.js
const debug = require('debug')('brig');
const fastq = require('fastq');
const workerpool = require('workerpool');

const WORKER_FILENAME = __dirname + '/worker.js';
const WORKER_OPTIONS = {
    minWorkers: process.env.MIN_WORKERS === 'max' ? 'max' : 1,
    maxQueueSize: process.env.MAX_QUEUE_SIZE === 'Infinity' ? Infinity : 255,
    workerType: process.env.WORKER_TYPE || 'auto',
};

function createQueue(worker, concurrency) {
    const queue = fastq(worker, concurrency);
    return queue;
}

function createWorkerPool(worker, options) {

    return workerpool.pool(worker, options);
}

function error(msg = 'Brig Unknown Error', throwErr = true) {
    // TODO: detect Error / String
    const err = new Error(msg);
    debug(err);
    // TODO: detect Promise / Function (callback) / false
    if (throwErr) throw err;
    return err;
}

class Brig {

    //private vars
    #closed = true;
    #queue = null;
    #workerpool = null;
    #timeout = process.env.WORKER_TIMEOUT === 'Infinity' ? Infinity : 1000 * 60;

    // private methods
    #exec(code, data) {
        return new Promise((respond, reject) => {
            debug('#exec(...)');

            this.#workerpool.exec('execute', [code, data])
                .timeout(this.#timeout)
                .then((resp) => {
                    debug('#exec(...) resp');
                    return respond(resp);
                })
                .catch((err) => {
                    return reject(err);
                });
        });
    };

    #queueWorker(arg, callback) {
        debug('#queueWorker(...) #exec');

        this.#exec(arg.code, arg.data)
            .then((resp) => {
                debug('#queueWorker(...) #exec resp');
                if (callback) callback(null, resp);

                // TODO: trigger response event
            })
            .catch((err) => {
                debug(err);

                error(err, !callback);
                if (callback) callback(err);
                // TODO: trigger err event
            });
    }

    // public methods
    // TODO: decide API
    send(code, data, callback) {
        debug('send(...)');

        if (this.#closed) {
            error('send fail: brig closed', !callback);
            if (callback) callback(err);
        }

        debug('send(...) queue push');
        this.#queue.push({code, data}, (err, resp) => {
            debug('send(...) queue push resp');
            return callback(err, resp);
        });

        return this;
    }

    open() {
        debug('open()');

        if (!this.#queue) {
            debug('open() create queue');
            //TODO: figure out queue concurrency
            this.#queue = createQueue(this, this.#queueWorker, 3);
        }
        else {
            if (this.#queue.resume) this.#queue.resume();
        }

        if (!this.#workerpool) {
            this.#workerpool = createWorkerPool(WORKER_FILENAME, WORKER_OPTIONS);
        }

        this.#closed = false;
        return this;
    }

    close(hard) {
        debug('close(%s)', !!hard);
        this.#closed = true;
        if (this.#queue) this.#queue.pause();
        if (hard) {
            if (this.#queue) this.#queue.killAndDrain ? this.#queue.killAndDrain() : this.#queue.kill();
            if (this.#workerpool) this.#workerpool.terminate();
        }
        else {
            // TODO: let queue try to drain naturally with a time limit
        }

        return this;
    }

    isOpen() {
        return !this.#closed;
    }

};

module.exports = Brig;